backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PoolPostgresExecutor.java

```diff
@@ -0,0 +1,107 @@
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.postgres.utils;
+
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.MAX_RETRY_ATTEMPTS;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.MIN_BACKOFF;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.PGSQL_DIALECT;
+import static org.apache.james.backends.postgres.utils.PostgresExecutor.SETTINGS;
+
+import java.time.Duration;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.inject.Inject;
+
+import org.apache.james.lifecycle.api.Disposable;
+import org.jooq.DSLContext;
+import org.jooq.Record;
+import org.jooq.impl.DSL;
+import org.reactivestreams.Publisher;
+
+import io.r2dbc.pool.ConnectionPool;
+import io.r2dbc.pool.ConnectionPoolConfiguration;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.R2dbcBadGrammarException;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
+
+public class PoolPostgresExecutor implements Disposable {
+
+    static int INITIAL_SIZE = 10;
+    static int MAX_SIZE = 50;
+    static Duration MAX_IDLE_TIME = Duration.ofMillis(5000);
+    private final ConnectionPool pool;
+
+    @Inject
+    public PoolPostgresExecutor(ConnectionFactory connectionFactory) {
+        ConnectionPoolConfiguration configuration = ConnectionPoolConfiguration.builder(connectionFactory)
+            .maxIdleTime(MAX_IDLE_TIME)
+            .initialSize(INITIAL_SIZE)
+            .maxSize(MAX_SIZE)
+            .build();
+        pool = new ConnectionPool(configuration);
+    }
+
+    public Mono<Void> executeVoid(Function<DSLContext, Mono<?>> queryFunction) {
+        return Mono.usingWhen(pool.create(),
+            connection -> dslContext(connection)
+                .flatMap(queryFunction)
+                .retryWhen(Retry.backoff(MAX_RETRY_ATTEMPTS, MIN_BACKOFF)
+                    .filter(preparedStatementConflictException()))
+                .then(),
+            Connection::close);
```

Comment:
I do not know the semantic of r2dbc pool but I do not like to see this create - close pattern.

I would prefer a semantic arround borrowing - releasing.

Reply:
This semantic was used as an example at README.md
https://github.com/r2dbc/r2dbc-pool